source/parser/x86_64/type_checker.hpp

```diff
@@ -27,12 +29,46 @@ namespace smeagle::x86_64 {
   inline bool is_ref(st::dataClass dc) { return dc == st::dataReference; }
 
   // Return true if dataClass is reference or pointer
-  inline bool is_indirect(st::dataClass dc) {
-    return dc == st::dataPointer || dc == st::dataReference;
-  }
+  inline bool is_indirect(st::dataClass dc) { return is_pointer(dc) || is_ref(dc); }
 
   // Return true if dataClass is a primitive type
-  inline bool is_primitive(st::dataClass dc) { return dc == st::dataEnum || dc == st::dataScalar; }
+  inline bool is_primitive(st::dataClass dc) { return dc == st::dataScalar; }
   inline bool is_typedef(st::dataClass dc) { return dc == st::dataTypedef; }
+  inline bool is_enum(st::dataClass dc) { return dc == st::dataEnum; }
+
+  // Dereference a pointer or reference
+  inline st::Type *deref(st::Type *t) {
+    if (is_pointer(t->getDataClass())) {
+      return deref(t->getPointerType()->getConstituentType());
+    }
+    if (is_ref(t->getDataClass())) {
+      return deref(t->getRefType()->getConstituentType());
+    }
+    return t;
+  }
+  // Unwrap and remove typedef
+  inline st::Type *remove_typedef(st::Type *t) {
+    if (is_typedef(t->getDataClass())) {
```

Comment:
Could this ever be further nested so we would want this to be a while?

Reply:
This should only remove one layer of typedef. In fact, it should probably be renamed to `remove_toplevel_typedef` or something. This is only needed because a reference might be hidden in the typedef. For example,

    struct Foo{};
    using foo_ref = Foo&;
    void foo(foo_ref x);

Calling `dedecorate` on `foo_ref` would lose the fact that there is a reference. I struggled with how to best split this behavior between here and `dedecorate`, and decided on this one. If you have some ideas for better usage, let's discuss!

Reply:
I think `remove_typedef` is well named - the "toplevel" could add to it but it's not necessary. So the name 'dedecorate" is based on the idea that pointers, references, and typedefs are decorations? Or is it that the underlying thing we want needs to be unwrapped? I'm not great at naming, but I think if we are trying to ultimately get to the underlying type (whatever it is) I would have:

- unwrap_underlying_type: implies unwrapping (removing) until you get the underlying type
- remove_typedef  : implies one removal

or something along those lines. We could probably figure out one way for the second function to use the first, but they are so simple it's probably not worth it.

Reply:
I changed its name a half-dozen times. I prefer unwrap_underlying_type! I'll change it and add comments.